### PR TITLE
refactor(search): Change datasets api search to have promise interfac…

### PR DESCRIPTION
**Pull request by markkopaas**
…e, to make it internally reusable.

### CODE REVIEW

File: applications/search/src/redux/modules/apis.js [hunk 1]
```diff
@@ -1,7 +1,10 @@
 import _ from 'lodash';
 import qs from 'qs';
 import { fetchActions } from '../fetchActions';
-import { apisSearchUrl } from '../../api/apis';
+import { apisSearch, apisSearchUrl, extractAggregations, extractApis, extractTotal } from '../../api/apis';
+import { reduxFsaThunk } from "../../lib/redux-fsa-thunk";
+import { datasetsSearch } from "../../api/datasets";
```
Comment (oyvindis) [inline applications/search/src/redux/modules/apis.js]: Sjekket lint? Jeg kan ikke se at denne importen er i bruk.

File: applications/search/src/redux/modules/apis.js [hunk 2]
```diff
@@ -1,7 +1,10 @@
...
+import { DATASETS_FAILURE, DATASETS_REQUEST, DATASETS_SUCCESS } from "./datasets";
```
Comment (oyvindis) [inline applications/search/src/redux/modules/apis.js]: Ikke i bruk.

File: applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx [hunk 3]
```diff
@@ -2,7 +2,11 @@ import _ from 'lodash';
 import { resolve } from 'react-resolver';
 import { InformationModelDetailsPage } from './information-model-details-page';
 import { getInformationmodel } from '../../api/informationmodels';
-import { getApiByHarvestSourceUri } from '../../api/apis';
+import { apisSearch, extractApis } from '../../api/apis';
+import noop from 'lodash/noop';
+
+export const getApiByHarvestSourceUri = harvestSourceUri =>
+  apisSearch({harvestSourceUri}).then(extractApis).catch(noop);
```
Comment (oyvindis) [inline applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx]: Den er fin!

**Review by oyvindis — APPROVED** [review]
Veldig bra! Bare noen import jeg tror kan fjernes.

File: applications/search/src/api/datasets.js [hunk 4]
```diff
@@ -16,3 +18,16 @@ export const getDatasetByURI = uri =>
     .get(`${datasetsUrlBase}/byuri?uri=${uri}`)
     .then(response => response.data)
     .catch(e => console.log(JSON.stringify(e))); // eslint-disable-line no-console
+
+export function extractDatasets(searchResponse) {
+  const hits = get(searchResponse, 'hits.hits', []);
+  return hits.map(hit => hit._source);
```
Comment (azaxarov) [inline applications/search/src/api/datasets.js]: The two lines could be done as one operation

File: applications/search/src/pages/concept-details-page/resolved-concept-details-page.jsx [hunk 5]
```diff
@@ -1,10 +1,18 @@
...
 import { resolve } from 'react-resolver';
 import { ConceptDetailsPage } from './concept-details-page';
-import { getConcept, getDatasets } from '../../api/concepts';
+import { getConcept } from '../../api/concepts';
+import { datasetsSearch, extractDatasets } from "../../api/datasets";
 
+const getDatasetsForConcept = (id)=> {
+  const subject = `https://fellesdatakatalog.brreg.no/api/concepts/${id}`;
+  const returnfields='id,uri,title';
+  return datasetsSearch({subject, returnfields})
```
Comment (azaxarov) [inline applications/search/src/pages/concept-details-page/resolved-concept-details-page.jsx]: Where does this function come from?

File: applications/search/src/redux/modules/apis.js [hunk 6]
```diff
@@ -48,9 +53,9 @@ export function apisReducer(state = initialState, action) {
     case APIS_SUCCESS:
       return {
         ...state,
-        apiItems: _.get(action.payload, 'hits'),
-        apiAggregations: _.get(action.payload, 'aggregations'),
-        apiTotal: _.get(action.payload, 'total'),
+        apiItems: extractApis(action.payload),
```
Comment (azaxarov) [inline applications/search/src/redux/modules/apis.js:53]: Is action.payload defined? It seems that meta is used in an action.

File: applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx (hunk 3 above)
Comment (azaxarov) [inline applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx]: How does this work? Does apiSearch take in variable key-value pairs?

File: applications/search/src/pages/datasets-report-page/report-stats/resolved-report-stats.jsx [hunk 7]
```diff
@@ -4,12 +4,12 @@ import { getDatasetStats } from '../../../api/get-dataset-stats';
...
 import { getApiStats } from '../../../api/get-api-stats';
 import { getConceptStats } from '../../../api/get-concept-stats';
-import { getConceptsByURIs } from '../../../api/concepts';
+import { conceptsSearch, extractConcepts } from '../../../api/concepts';
 
 const memoizedGetDatasetStats = _.memoize(getDatasetStats);
 const memoizedGetApiStats = _.memoize(getApiStats);
 const memoizedGetConceptStats = _.memoize(getConceptStats);
-const memoizedGetConceptsByURIs = _.memoize(getConceptsByURIs);
+const memoizedSearchConcepts = _.memoize(conceptsSearch, JSON.stringify);
```
Comment (azaxarov) [inline applications/search/src/pages/datasets-report-page/report-stats/resolved-report-stats.jsx:12]: What is the purpose of JSON.stringify?

File: applications/search/src/redux/modules/informationModels.js [hunk 8]
```diff
@@ -1,8 +1,13 @@
 import _ from 'lodash';
 import qs from 'qs';
```
Comment (azaxarov) [inline applications/search/src/redux/modules/informationModels.js]: Probably not in use

File: applications/search/src/redux/modules/informationModels.js [hunk 9]
```diff
@@ -50,13 +55,9 @@ export function informationModelsReducer(state = initialState, action) {
...
-        informationModelItems: _.get(
-          action.payload,
-          '_embedded.informationmodels'
-        ),
-        informationModelAggregations: _.get(action.payload, 'aggregations'),
-        informationModelTotal: _.get(action.payload, ['page', 'totalElements']),
-
+        informationModelItems: extractInformationmodels(action.payload),
+        informationModelAggregations: extractAggregations(action.payload),
+        informationModelTotal: extractTotal(action.payload),
```
Comment (azaxarov) [inline applications/search/src/redux/modules/informationModels.js:57]: Is action.payload ok to use?

**Review by azaxarov — CHANGES_REQUESTED** [review]

File: applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx (hunk 3 above)
Comment (markkopaas) [reply applications/search/src/pages/informationmodel-details-page/resolved-information-model-details-page.jsx]: takk

File: applications/search/src/redux/modules/informationModels.js (hunk 8 above)
Comment (markkopaas) [reply applications/search/src/redux/modules/informationModels.js]: lint

File: applications/search/src/api/informationmodels.js [hunk 10]
```diff
@@ -1,5 +1,4 @@
 import axios from 'axios';
-import _ from 'lodash';
 import get from "lodash/get";
```
Comment (markkopaas) [inline applications/search/src/api/informationmodels.js]: lint

File: applications/search/src/api/informationmodels.js [hunk 11]
```diff
@@ -1,5 +1,4 @@
...
 
 const informationmodelsUrlBase = `/api/informationmodels`;
```
Comment (markkopaas) [inline applications/search/src/api/informationmodels.js:4]: ren string

File: applications/search/src/api/datasets.js (hunk 4 above)
Comment (markkopaas) [reply applications/search/src/api/datasets.js]: fikset

File: applications/search/src/redux/modules/informationModels.js (hunk 8 above)
Comment (markkopaas) [reply applications/search/src/redux/modules/informationModels.js]: det er i bruk til å generere en kompakt nøkkel. JSON.stringify kunne også funke, menn det gir litt lengere nøkkel... skal vi fikse?

File: applications/search/src/redux/modules/informationModels.js [hunk 12]
```diff
@@ -1,14 +1,18 @@
...
+  informationmodelsSearch
+} from '../../api/informationmodels';
+import { reduxFsaThunk } from '../../lib/redux-fsa-thunk';
 
 export const INFORMATIONMODELS_REQUEST = 'INFORMATIONMODELS_REQUEST';
 export const INFORMATIONMODELS_SUCCESS = 'INFORMATIONMODELS_SUCCESS';
 export const INFORMATIONMODELS_FAILURE = 'INFORMATIONMODELS_FAILURE';
 
-const generateQueryKey = query => qs.stringify(query, { skipNulls: true });
+const generateQueryKey = query => JSON.stringify(query);
```
Comment (azaxarov) [inline applications/search/src/redux/modules/informationModels.js]: Function is not necessary here, Json.stringify can be moved it to line 28.

**Review by azaxarov — APPROVED** [review]
Just a small comment

**Pull request merged**